lib/posix/pthread.c

```diff
@@ -216,18 +216,18 @@ int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
  */
 int pthread_setcancelstate(int state, int *oldstate)
 {
+	k_spinlock_key_t cancel_key;
 	struct posix_thread *pthread = to_posix_thread(pthread_self());
 
 	if (state != PTHREAD_CANCEL_ENABLE &&
 	    state != PTHREAD_CANCEL_DISABLE) {
 		return EINVAL;
 	}
 
+	cancel_key = k_spin_lock(&pthread->cancel_lock);
 	*oldstate = pthread->cancel_state;
-
-	pthread_mutex_lock(&pthread->cancel_lock);
 	pthread->cancel_state = state;
-	pthread_mutex_unlock(&pthread->cancel_lock);
+	k_spin_unlock(&pthread->cancel_lock, cancel_key);
 
 	if (state == PTHREAD_CANCEL_ENABLE && pthread->cancel_pending) {
```

Comment:
Should the fetch of `cancel_pending` be done under the lock? It seems easier to understand the locking that way with all access to `cancel_state` and `cancel_pending` done under the lock?

Reply:
Yes - I believe that was actually a bug.